messaginginapp/src/main/java/io/customer/messaginginapp/state/InAppMessagingState.kt

```diff
@@ -32,19 +32,53 @@ data class InAppMessagingState(
 }
 
 sealed class MessageState {
-    object Default : MessageState()
-    data class Processing(val message: Message) : MessageState()
-    data class Loaded(val message: Message) : MessageState()
+    object Initial : MessageState()
+    data class Loading(val message: Message) : MessageState()
+    data class Displayed(val message: Message) : MessageState()
     data class Embedded(val message: Message, val elementId: String) : MessageState()
     data class Dismissed(val message: Message) : MessageState()
 
     override fun toString(): String {
         return when (this) {
-            is Default -> "Default"
-            is Processing -> "Processing(message=${message.queueId})"
-            is Loaded -> "Loaded(message=${message.queueId})"
+            is Initial -> "Initial"
+            is Loading -> "Loading(message=${message.queueId})"
+            is Displayed -> "Displayed(message=${message.queueId})"
             is Embedded -> "Embedded(message=${message.queueId}, elementId=$elementId)"
             is Dismissed -> "Dismissed(message=${message.queueId})"
         }
     }
 }
+
+fun InAppMessagingState.diff(other: InAppMessagingState): Map<String, Pair<Any?, Any?>> {
+    val differences = mutableMapOf<String, Pair<Any?, Any?>>()
+
+    if (this.siteId != other.siteId) {
+        differences["siteId"] = Pair(this.siteId, other.siteId)
+    }
+    if (this.dataCenter != other.dataCenter) {
+        differences["dataCenter"] = Pair(this.dataCenter, other.dataCenter)
+    }
+    if (this.environment != other.environment) {
+        differences["environment"] = Pair(this.environment, other.environment)
+    }
+    if (this.pollInterval != other.pollInterval) {
+        differences["pollInterval"] = Pair(this.pollInterval, other.pollInterval)
+    }
+    if (this.userId != other.userId) {
+        differences["userId"] = Pair(this.userId, other.userId)
+    }
+    if (this.currentRoute != other.currentRoute) {
+        differences["currentRoute"] = Pair(this.currentRoute, other.currentRoute)
+    }
+    if (this.currentMessageState != other.currentMessageState) {
+        differences["currentMessageState"] = Pair(this.currentMessageState, other.currentMessageState)
+    }
+    if (this.messagesInQueue != other.messagesInQueue) {
+        differences["messagesInQueue"] = Pair(this.messagesInQueue, other.messagesInQueue)
+    }
+    if (this.shownMessageQueueIds != other.shownMessageQueueIds) {
+        differences["shownMessageQueueIds"] = Pair(this.shownMessageQueueIds, other.shownMessageQueueIds)
+    }
+
+    return differences
+}
```

Comment:
Since the comparison is mostly same, we can eliminate repetitive code and implement something like this:

```kotlin
fun InAppMessagingState.diff(other: InAppMessagingState): Map<String, Pair<Any?, Any?>> {
    val differences = mutableMapOf<String, Pair<Any?, Any?>>()

    fun <T> putIfDifferent(key: String, value: InAppMessagingState.() -> T) {
        val thisValue = value(this)
        val otherValue = value(other)
        if (thisValue != otherValue) {
            differences[key] = Pair(thisValue, otherValue)
        }
    }

    putIfDifferent("siteId") { siteId }
    putIfDifferent("environment") { environment }
    // other values

    return differences
}
```

We can use `Any` instead of `T` too to further simplify if needed.

Reply:
I think it can be simplified even more than that, I wanted to keep it readable 😄 let me find another way